chips/nrf52/src/i2c.rs

```diff
@@ -0,0 +1,446 @@
+//! Implementation of I2C for nRF52 using EasyDMA.
+//!
+//! This module supports nRF52's two I2C master (`TWIM`) peripherals,
+//! but not I2C slave (`TWIS`).
+//!
+//! - Author: Jay Kickliter
+//! - Author: Andrew Thompson
+//! - Date: Nov 4, 2017
+
+use core::cell::Cell;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil;
+use nrf5x::nvic;
+use nrf5x::peripheral_interrupts::NvicIdx;
+use nrf5x::pinmux::Pinmux;
+
+/// An I2C master device.
+///
+/// A `TWIM` instance wraps a `registers::TWIM` together with
+/// addition data necessary to implement an asynchronous interface.
+pub struct TWIM {
+    registers: *const registers::TWIM,
+    nvic_idx: NvicIdx,
+    client: Cell<Option<&'static hil::i2c::I2CHwMasterClient>>,
+    busy: Cell<bool>,
+    buf: TakeCell<'static, [u8]>,
+    enabled: Cell<bool>,
+}
+
+impl TWIM {
+    const fn new(instance: usize) -> TWIM {
+        TWIM {
+            registers: registers::INSTANCES[instance].0,
+            nvic_idx: registers::INSTANCES[instance].1,
+            client: Cell::new(None),
+            busy: Cell::new(false),
+            buf: TakeCell::empty(),
+            enabled: Cell::new(false),
+        }
+    }
+
+    pub fn set_client(&self, client: &'static hil::i2c::I2CHwMasterClient) {
+        debug_assert!(self.client.get().is_none());
+        self.client.set(Some(client));
+    }
+
+    fn regs(&self) -> &registers::TWIM {
+        unsafe { &*self.registers }
+    }
+
+    fn enable_nvic(&self) {
+        nvic::enable(self.nvic_idx);
+    }
+
+    /// Configures an already constructed `TWIM`.
+    pub fn configure(&self, scl: Pinmux, sda: Pinmux) {
+        let regs = self.regs();
+        regs.psel_scl.set(scl);
+        regs.psel_sda.set(sda);
+        self.enable_nvic();
+    }
+
+    /// Enables hardware TWIM peripheral.
+    pub fn enable(&self) {
+        let regs = self.regs();
+        regs.enable.set(registers::Enable::Enabled);
+        self.enabled.set(true);
+    }
+
+    /// Disables hardware TWIM peripheral.
+    pub fn disable(&self) {
+        let regs = self.regs();
+        regs.enable.set(registers::Enable::Disabled);
+        self.enabled.set(false);
+    }
+
+    pub fn handle_interrupt(&self) {
+        if self.regs().events_stopped.get() == 1 {
+            self.regs().events_stopped.set(0);
+            match self.client.get() {
```

Comment:
Looking like a good place for the new `?` support for `Option`. Unfortunately, I think we missed the the nightly that added it in #690 by a couple weeks. We can try it out next time...